### ISSUE TITLE
docs: recommendations regarding blocs adding internal events

### ISSUE THREAD

**Issue by felangel** [felangel/bloc#3594]
**Description**

As a developer I want to have documentation that describes if/when a bloc should add internal events so that I can have a better understanding of best-practices, potential pitfalls, etc.

This can probably be part of the FAQs section of the bloclibrary.dev docs.

**Comment by KyleFin** [comment]
Thanks for filing! A few thoughts:

Diagram at the top of https://bloclibrary.dev/#/architecture shows `events` are for UI to send information to bloc, and examples I found throughout bloclibrary.dev tutorials generally show `.add(MyEvent)` being called in UI code, so I wasn't sure if adding events in an event handler is discouraged or just not the common case.

Main pitfall I encountered was event being added after async work completed (bloc was closed after the async work started).

Something like this:

```
Future<void> _handleMyEvent(MyEvent e, Emitter<MyState> emit) async {
  emit(MyState('state on receiving event'));
  final result = await repository.fetchData();    // User can navigate to a different screen after fetch started and bloc is closed
  add(MyPostProcessingEvent());    // Throws if bloc is closed
}
```

Also, from our discussion it might be helpful to explain a bit about how/when `emit.onEach` or `emit.forEach` can make this safer.
